src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+
+namespace Microsoft.CodeAnalysis.AddDebuggerDisplay
+{
+    internal abstract class AbstractAddDebuggerDisplayCodeRefactoringProvider<TTypeDeclarationSyntax, TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_trackingAnnotation = new SyntaxAnnotation();
+
+        private const string DebuggerDisplayMethodName = "GetDebuggerDisplay";
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var type =
+                await context.TryGetRelevantNodeAsync<TTypeDeclarationSyntax>().ConfigureAwait(false)
+                ?? await GetRelevantTypeFromMethodAsync(context).ConfigureAwait(false);
+
+            if (type is null) return;
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var typeSymbol = (ITypeSymbol)semanticModel.GetDeclaredSymbol(type, context.CancellationToken);
+
+            if (IsClassOrStruct(typeSymbol) && !HasDebuggerDisplayAttribute(typeSymbol, semanticModel.Compilation))
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Add_DebuggerDisplay,
+                    cancellationToken => ApplyAsync(context.Document, type, cancellationToken)));
+            }
+        }
+
+        private static async Task<TTypeDeclarationSyntax> GetRelevantTypeFromMethodAsync(CodeRefactoringContext context)
+        {
+            var method = await context.TryGetRelevantNodeAsync<TMethodDeclarationSyntax>().ConfigureAwait(false);
+            if (method != null)
+            {
+                var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+                var methodSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(method);
+
+                if (IsToStringOverride(methodSymbol) || IsDebuggerDisplayMethod(methodSymbol))
+                {
+                    return method.FirstAncestorOrSelf<TTypeDeclarationSyntax>();
+                }
+            }
+
+            return null;
+        }
+
+        private static bool IsToStringOverride(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol is
+            {
+                Arity: 0,
+                Parameters: { IsEmpty: true },
+                Name: nameof(ToString),
+                IsOverride: true
+            };
+        }
+
+        private static bool IsDebuggerDisplayMethod(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol is
+            {
+                Arity: 0,
+                Parameters: { IsEmpty: true },
+                Name: DebuggerDisplayMethodName
+            };
+        }
+
+        private static bool IsClassOrStruct(ITypeSymbol typeSymbol)
```

Comment:
nit `=>` and switch-expression.

Reply:
Switch expression is better? I was on the fence. No-brainer for me once we have `or` patterns.

Reply:
hnoestly, i'd just write this as `typeSymbol.TypeKind == Class || typeSymbol.TypeKind == Struct`.